### ISSUE TITLE
Server operation handlers need to be able to result with internal error

### ISSUE THREAD

**Issue by guymguym** [awslabs/smithy-rs#1150]
Servers might have internal errors which are not defined in the API as functional errors.
However, currently if the operation has no errors defined in the API, the handler type will not allow to return any error, and the return type of the handler will be expected to be the output type directly, which makes it impossible to signal an internal error has occurred. Only panic is possible which is not desired for most servers.
Is there's any way to map such errors in the server which does not panic?
Maybe adding an `unhandled` error type like the client has?

That said - when I look at the S3 smithy model, I find that it lacks a lot of error definitions for operations.
Here is a list of all the operations that have errors defined:
```
$ node -e 'm=require("../smithy-rs/aws/sdk/aws-models/s3.json"); Object.entries(m.shapes).forEach(([n,s])=>{ if (s.type==="operation" && s.errors) console.log(n, "---", s.errors.map(x=>x.target).join(",")); })'
com.amazonaws.s3#AbortMultipartUpload --- com.amazonaws.s3#NoSuchUpload
com.amazonaws.s3#CopyObject --- com.amazonaws.s3#ObjectNotInActiveTierError
com.amazonaws.s3#CreateBucket --- com.amazonaws.s3#BucketAlreadyExists,com.amazonaws.s3#BucketAlreadyOwnedByYou
com.amazonaws.s3#GetObject --- com.amazonaws.s3#InvalidObjectState,com.amazonaws.s3#NoSuchKey
com.amazonaws.s3#GetObjectAcl --- com.amazonaws.s3#NoSuchKey
com.amazonaws.s3#HeadBucket --- com.amazonaws.s3#NotFound
com.amazonaws.s3#HeadObject --- com.amazonaws.s3#NotFound
com.amazonaws.s3#ListObjects --- com.amazonaws.s3#NoSuchBucket
com.amazonaws.s3#ListObjectsV2 --- com.amazonaws.s3#NoSuchBucket
com.amazonaws.s3#PutObjectAcl --- com.amazonaws.s3#NoSuchKey
com.amazonaws.s3#RestoreObject --- com.amazonaws.s3#ObjectAlreadyInActiveTierError
```
This is clearly a tiny fraction of the real errors that this API can return.
See https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html

I think that I could bypass this limitation if the server handlers could return a generic error in addition to the predefined ones...
@david-perez @crisidev WDYT?

**Comment by crisidev** [comment]
For a good API definition, if an operation doesn't allow errors, it should be generated as infallible as we are doing right now. IMO it is the user responsibility to handle this, depending on their design.

What I am doing for this in a side project I am working on with smithy-rs is: have a fallible method that wraps all the logic of the operation and just call something like 

```rust
.unwrap_or_else(|e| {
    error!("This should not happen: {}", e);
    MyOutput::builder().build()
})
```

inside the actual operation handler. 

Long story short I believe what we are doing is the right thing, but I am happy to hear thoughts from others..

**Comment by guymguym** [comment]
It is correct, if the API is always complete, but in real life it seems that the API is always incomplete.
An "empty" output is not always possible, and in many cases can be more confusing than an unexpected error that the client can at least read.
In addition, there are network errors, which can always cause even the best server to fail.

I guess that even if you wouldn't want most servers to use this option - it would be practically cumbersome to require that all servers will commit to having no unexpected errors, and currently the only option is to panic, which is not a real option.

**Comment by crisidev** [comment]
If we are specifically talking about the S3 model you are using, I can see why it is a problem for you, but we should fix that model instead of building workarounds to cope with the fact that it is incomplete.

**Comment by guymguym** [comment]
Perhaps fixing the S3 model is possible, but I'm not sure in what timeframe. Even if so, I still think that it's not flexible enough and IMHO real life server handlers should aways have that option to introduce an unexpected failure for a request which was not known to their clients on codegen time.

**Comment by crisidev** [comment]
After talking a little with @rcoh, a good middleground solution would to build a Decorator specific for the legacy models like S3 one, which can be applied during codegen and for example make all operation fallible, no matter what the model describes. 

It is already something happening in the client codegen, like [here](https://github.com/awslabs/smithy-rs/blob/2e7ed943513203f1472f2490866dc4fb8a392bd3/aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/customize/s3/S3Decorator.kt).

This would give us the freedom to plug the decorator if needed and will prevent workaround in the codegen for users that have the luxury to work with new models that adheres properly to the specs. WDYT?

**Comment by guymguym** [comment]
Sounds great @crisidev! That's exactly the type of flexibility I was looking for.

**Comment by crisidev** [comment]
Happy to hear that. Now the issue is that we do not have anything really ready to use Decorators nor the Decorator itself in the server codegen, it is the first time someone is asking for it.

I can take a closer look to this next week it that's ok with you..

**Comment by guymguym** [comment]
That would be very appreciated. Let me know how I can help.
In the meanwhile, I'm just panic-ing on error :)

**Comment by guymguym** [comment]
Another question @crisidev - Where should I check for the model completeness?
Should I open an issue in https://github.com/awslabs/smithy ?

**Comment by crisidev** [comment]
What do you mean with model completeness?

**Comment by crisidev** [comment]
We did a little bit of research and it looks like we can implement a decorator implementing [transformModel](https://github.com/awslabs/smithy-rs/blob/main/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/customize/RustCodegenDecorator.kt#L70). 

`transformModel` will add a new error `InternalServerError` (or something like it) to all operations that are infallible and let the codegen generate everything.

**Comment by guymguym** [comment]
> What do you mean with model completeness?

I mean that these api errors are not defined in the smithy s3 model - https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
